File: src/registry/sut-registry.ts
/**
 * SUT Registry
 *
 * Central registry for System Under Test definitions. SUTs are registered
 * with their factories and metadata, enabling lazy instantiation during
 * experiment execution.
 */

import type { SUT, SutDefinition, SutFactory, SutRegistration, SutRole } from "../types/sut.js";

/**
 * Registry for System Under Test definitions.
 *
 * @template TInputs - The algorithm inputs type
 * @template TResult - The algorithm result type
 */
export class SUTRegistry<TInputs = unknown, TResult = unknown> {
	private readonly definitions = new Map<string, SutDefinition<TInputs, TResult>>();

	/**
	 * Register a new SUT.
	 *
	 * @param registration - SUT metadata
	 * @param factory - Factory for creating SUT instances
	 * @throws Error if SUT with same ID already registered
	 */
	register(registration: SutRegistration, factory: SutFactory<TInputs, TResult>): this {
		if (this.definitions.has(registration.id)) {
			throw new Error(`SUT already registered: ${registration.id}`);
		}

		this.definitions.set(registration.id, { registration, factory });
		return this;
	}

	/**
	 * Get a SUT definition by ID.
	 *
	 * @param id - SUT identifier
	 * @returns SUT definition or undefined
	 */
	get(id: string): SutDefinition<TInputs, TResult> | undefined {
		return this.definitions.get(id);
	}

	/**
	 * Get a SUT definition by ID, throwing if not found.
	 *
	 * @param id - SUT identifier
	 * @returns SUT definition
	 * @throws Error if SUT not found
	 */
	getOrThrow(id: string): SutDefinition<TInputs, TResult> {
		const definition = this.definitions.get(id);
		if (!definition) {
			throw new Error(`SUT not found: ${id}`);
		}
		return definition;
	}

	/**
	 * Get all SUTs with a specific role.
	 *
	 * @param role - Role to filter by
	 * @returns Array of matching SUT definitions
	 */
	getByRole(role: SutRole): SutDefinition<TInputs, TResult>[] {
		return [...this.definitions.values()].filter((d) => d.registration.role === role);
	}

	/**
	 * Get all SUTs with a specific tag.
	 *
	 * @param tag - Tag to filter by
	 * @returns Array of matching SUT definitions
	 */
	getByTag(tag: string): SutDefinition<TInputs, TResult>[] {
		return [...this.definitions.values()].filter((d) => d.registration.tags.includes(tag));
	}

	/**
	 * List all registered SUT IDs.
	 *
	 * @returns Array of SUT identifiers
	 */
	list(): string[] {
		return [...this.definitions.keys()];
	}

	/**
	 * List all registered SUT registrations.
	 *
	 * @returns Array of SUT registrations
	 */
	listRegistrations(): SutRegistration[] {
		return [...this.definitions.values()].map((d) => d.registration);
	}

	/**
	 * Check if a SUT is registered.
	 *
	 * @param id - SUT identifier
	 * @returns true if registered
	 */
	has(id: string): boolean {
		return this.definitions.has(id);
	}

	/**
	 * Get the number of registered SUTs.
	 */
	get size(): number {
		return this.definitions.size;
	}

	/**
	 * Clear all registrations.
	 */
	clear(): void {
		this.definitions.clear();
	}

	/**
	 * Create a new SUT instance.
	 *
	 * @param id - SUT identifier
	 * @param config - Optional configuration overrides
	 * @returns SUT instance ready for execution
	 */
	create(id: string, config?: Record<string, unknown>): SUT<TInputs, TResult> {
		const definition = this.getOrThrow(id);
		return definition.factory(config);
	}
}

/**
 * Global SUT registry instance.
 *
 * Use this for standard registration, or create instances for isolation.
 */
export const sutRegistry = new SUTRegistry();

File: python/src/ppef/registry/sut_registry.py
"""SUT Registry.

Central registry for System Under Test definitions. SUTs are registered
with their factories and metadata, enabling lazy instantiation during
experiment execution.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, TypeVar

from ppef.types.sut import SutRegistration, SutRole

TInputs = TypeVar("TInputs")
TResult = TypeVar("TResult")


class SUT(Protocol[TInputs, TResult]):
    """Protocol for a System Under Test instance."""

    @property
    def id(self) -> str: ...

    @property
    def config(self) -> dict[str, object]: ...

    async def run(self, inputs: TInputs) -> TResult: ...


class SutFactory(Protocol[TInputs, TResult]):
    """Protocol for a SUT factory callable."""

    def __call__(self, config: dict[str, object] | None = None) -> SUT[TInputs, TResult]: ...


@dataclass(frozen=True)
class SutDefinition[TInputs, TResult]:
    """A registered SUT definition pairing metadata with a factory."""

    registration: SutRegistration
    factory: SutFactory[TInputs, TResult]


class SUTRegistry[TInputs, TResult]:
    """Registry for System Under Test definitions.

    Type Parameters:
        TInputs: The algorithm inputs type.
        TResult: The algorithm result type.
    """

    def __init__(self) -> None:
        self._definitions: dict[str, SutDefinition[TInputs, TResult]] = {}

    def register(
        self,
        registration: SutRegistration,
        factory: SutFactory[TInputs, TResult],
    ) -> SUTRegistry[TInputs, TResult]:
        """Register a new SUT.

        Args:
            registration: SUT metadata.
            factory: Factory for creating SUT instances.

        Returns:
            Self for method chaining.

        Raises:
            ValueError: If a SUT with the same ID is already registered.
        """
        if registration.id in self._definitions:
            raise ValueError(f"SUT already registered: {registration.id}")

        self._definitions[registration.id] = SutDefinition(
            registration=registration,
            factory=factory,
        )
        return self

    def get(self, id: str) -> SutDefinition[TInputs, TResult] | None:
        """Get a SUT definition by ID.

        Args:
            id: SUT identifier.

        Returns:
            SUT definition or None if not found.
        """
        return self._definitions.get(id)

    def get_or_raise(self, id: str) -> SutDefinition[TInputs, TResult]:
        """Get a SUT definition by ID, raising if not found.

        Args:
            id: SUT identifier.

        Returns:
            SUT definition.

        Raises:
            KeyError: If the SUT is not found.
        """
        definition = self._definitions.get(id)
        if definition is None:
            raise KeyError(f"SUT not found: {id}")
        return definition

    def get_by_role(self, role: SutRole) -> list[SutDefinition[TInputs, TResult]]:
        """Get all SUTs with a specific role.

        Args:
            role: Role to filter by.

        Returns:
            List of matching SUT definitions.
        """
        return [d for d in self._definitions.values() if d.registration.role == role]

    def get_by_tag(self, tag: str) -> list[SutDefinition[TInputs, TResult]]:
        """Get all SUTs with a specific tag.

        Args:
            tag: Tag to filter by.

        Returns:
            List of matching SUT definitions.
        """
        return [d for d in self._definitions.values() if tag in d.registration.tags]

    def list_all(self) -> list[str]:
        """List all registered SUT IDs.

        Returns:
            List of SUT identifiers.
        """
        return list(self._definitions.keys())

    def list_registrations(self) -> list[SutRegistration]:
        """List all registered SUT registrations.

        Returns:
            List of SUT registrations.
        """
        return [d.registration for d in self._definitions.values()]

    def has(self, id: str) -> bool:
        """Check if a SUT is registered.

        Args:
            id: SUT identifier.

        Returns:
            True if registered.
        """
        return id in self._definitions

    @property
    def size(self) -> int:
        """Get the number of registered SUTs."""
        return len(self._definitions)

    def clear(self) -> None:
        """Clear all registrations."""
        self._definitions.clear()

    def create(
        self,
        id: str,
        config: dict[str, object] | None = None,
    ) -> SUT[TInputs, TResult]:
        """Create a new SUT instance.

        Args:
            id: SUT identifier.
            config: Optional configuration overrides.

        Returns:
            SUT instance ready for execution.
        """
        definition = self.get_or_raise(id)
        return definition.factory(config)


sut_registry: SUTRegistry[object, object] = SUTRegistry()
"""Global SUT registry instance.

Use this for standard registration, or create instances for isolation.
"""
